Create Editor Services architecture

wouterjsmit opened the pull request:
Implements an architecture design for the Editor Servicese.

- Editor services is an object that `IRepl` uses to request specific `FailOrSuccessResults<>` of some service result.
- The Services interface directly with `FailableFunctions`.
- The functions are composed by `FunctionComposer`, which is a similar factory/builder to `CommandBuilder<B>`. The Functioncomposer takes over some responsibilities of the commandbuilder and decouples the function composition from the `IReplCommand`s.
- The Eclipse frontend implements code to enable to display of highlighted regions.
- The clients get a `FailOrSuccessVisitor`, which can only visit `FailOrSuccessResult`s (These are still visitable in the original way as well, as part of the `IResult` interface).
- The facilitate the dispatch of exceptions in the return (`ExceptionResult`), a third option to the FailOrSuccess data type was added. Whether this should stay is debatable, but currently it is safer to not break original behaviour.

wouterjsmit committed:
Add Editor Services interface

Provides Editor services that directly invoke the FailableFunctions by
using the FunctionComposer. The Editor Services can be injected using
Guice, to the frontend and used in a similar way to the CommandInvoker.
The presence of a separate class for this, is that the exact result type
can be determined, while with IReplCommands this was not the case.

The Editor uses a strategy pattern to define its behaviour before and
after loading a language. This is necessary to provide a clean interface
to the client, who must receive its EditorServices object from the
start, even though a language is not loaded initially, so that it
doesn't need to do null checks all the time.
The strategy of the EditorServices will decide whether it can provide a
meaningful result or simply return an ExceptionResult.

wouterjsmit committed:
Add Excepted option in FailOrSuccessResult

The third options enables dispatching in an either visitor to the
exception result.

wouterjsmit committed:
Create FailOrSuccessVisitor

This visitor only visits FailOrSuccess types (Success,Failed,Excepted).

wouterjsmit committed:
Create SyntaxVisitor in front-end

Create a syntaxvisitor that is used in the frontend to control the
codeflow of a syntax highlighting request.

wouterjsmit committed:
Create observer for input field keypresses

This event must trigger the syntax highlighting.

wouterjsmit committed:
Create Style failable function

The Style function adds interfacing with Spoofax to receive syntax
highlighting.

The Style function is a FailbleFunction and can be invoked in the same
way.

wouterjsmit committed:
Use Editor Services in front end

wouterjsmit committed:
Support highlighted Style Regions in Eclipse front

These will be called when a syntax highlighting result is received.

justinvdk commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/services/ServiceUnavailableException.java:
```diff
@@ -0,0 +1,17 @@
+
+public class ServiceUnavailableException extends Exception {
+
+    private static final long serialVersionUID = 1L;
```
Shouldn't this be some generated value?

wouterjsmit replied:
Suggestions? Eclipse quick fix gave this.

justinvdk commented on org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/services/ServiceUnavailableException.java:
```diff
@@ -0,0 +1,17 @@
+package org.metaborg.spoofax.shell.services;
+
+public class ServiceUnavailableException extends Exception {
```
Extend from MetaborgException perhaps? (Only if it makes sense..)

wouterjsmit replied:
It is not a `MetaBorg` exception (that is the Spoofax logic). Could create a super type for Shell exceptions, but currently the existing `CommandNotFoundException` simply extends from `Exception` as well.
Only benefit would be to catch Shell exceptions in one nominator.

justinvdk commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/EclipseEditor.java:
```diff
@@ -76,9 +89,14 @@ public void setFocus() {
-    public Observable<String> asObservable() {
-        return Observable
-            .create((Observable.OnSubscribe<String>) EclipseEditor.this.observers::add);
+    public Observable<String> asObservable(boolean live) {
```
Discuss:
Is this the right way? Will there ever be more visitable aspects to this class?

wouterjsmit replied:
I don't think there should be, but we could consider a framework for the client. Currently the only framework is that the client **must** implement an `IRepl` for the endpoint and an `IResultVisitor` to display results.

If the back end requires knowledge of an input component, or we want to provide a framework for that component to live in (like `IDisplay` for `IResultVisitor` is a framework for the output window), we could consider specifying an interface for "`IReplEditor`" as well. See also the createRepl factory method (`IWidgetFactory`).

justinvdk commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/IWidgetFactory.java:
```diff
@@ -29,12 +29,14 @@
-     * @return The created {@link EclipseRepl}.
-     */
-    EclipseRepl createRepl(IDisplay display);
+	 * Instantiate a new {@link EclipseRepl}.
```
Indentation is inconsistent

wouterjsmit committed:
Split FunctionComposer from CommandBuilder

The commandbuilder dealt with both the function compositions (in terms
of private methods) and building commands.

The composed functions need to be exposed to be called by editor
services directly, rather than via the IReplCommand invocation.
This is important because the IReplCommand loses the return type
information. The exact expected type is important to the client, who
must use that result to implement the editor feature.

This change adds a FunctionComposer, simply specifying how to compose
functions.

It can be instantiated in the same way as the CommandBuilder: using the
factory method in IFunctionFactory.

The commandBuilder uses the composer to build the commands with, rather
than doing it himself.

wouterjsmit committed:
Use Editor Services in front end
